integrationTesting/tests/organize/surveys/submitting-survey.spec.ts

```diff
@@ -39,6 +40,84 @@ test.describe('User submitting a survey', () => {
     moxy.teardown();
   });
 
+  test('submits text input', async ({ appUri, moxy, page }) => {
+    await page.goto(
+      `${appUri}/o/${KPDMembershipSurvey.organization.id}/surveys/${KPDMembershipSurvey.id}`
+    );
+
+    await page.fill('[name="2.text"]', 'Topple capitalism');
+    await page.click('input[name="sig"][value="anonymous"]');
+    await page.click('data-testid=Survey-acceptTerms');
+    await Promise.all([
+      page.waitForResponse((res) => res.request().method() == 'POST'),
+      await page.click('data-testid=Survey-submit'),
+    ]);
+
+    const log = moxy.log(`/v1${apiPostPath}`);
+    expect(log.length).toBe(1);
+    const [request] = log;
+    const data = request.data as {
+      responses: ZetkinSurveyQuestionResponse[];
+    };
+    expect(data.responses).toMatchObject([
+      {
+        question_id: KPDMembershipSurvey.elements[1].id,
+        response: 'Topple capitalism',
+      },
+    ]);
+  });
+
+  test('required text input blocks submission when empty', async ({
+    appUri,
+    moxy,
+    page,
+  }) => {
+    moxy.setZetkinApiMock(
+      `/orgs/${KPDMembershipSurvey.organization.id}/surveys/${KPDMembershipSurvey.id}`,
+      'get',
+      {
+        ...KPDMembershipSurvey,
+        elements: KPDMembershipSurvey.elements.map((element, i) => {
+          if (i === 1) {
+            return {
+              ...element,
+              question: {
+                ...(element as ZetkinSurveyQuestionElement).question,
+                required: true,
+              },
+            };
+          }
+          return element;
+        }),
+      }
+    );
+
+    await page.goto(
+      `${appUri}/o/${KPDMembershipSurvey.organization.id}/surveys/${KPDMembershipSurvey.id}`
+    );
+
+    moxy.setZetkinApiMock(
+      `/orgs/${KPDMembershipSurvey.organization.id}/surveys/${KPDMembershipSurvey.id}/submissions`,
+      'post',
+      {
+        timestamp: '1857-05-07T13:37:00.000Z',
+      }
+    );
+
+    const requiredTextInput = await page.locator('[name="2.text"][required]');
+    await requiredTextInput.waitFor({ state: 'visible' });
+
+    await page.click('input[name="sig"][value="anonymous"]');
+    await page.click('data-testid=Survey-acceptTerms');
+    await page.click('data-testid=Survey-submit');
+
+    const validationMessage = await requiredTextInput.evaluate((element) => {
+      const input = element as HTMLTextAreaElement;
+      return input.validationMessage;
+    });
+    expect(validationMessage).toBe('Please fill in this field.');
```

Comment:
Where is this text coming from? Is that the browser's message, or a message we created?

Reply:
this is browser-generated. might be better to use [validity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/validity) here instead tbh

Reply:
Got it! I'm guessing the browser message could theoretically change between browser versions, which would cause the test to fail even if validation hasn't changed. So testing for something more to-the-point might make sense.